### DISCUSSION TITLE
Qieyun.js 0.13 更動概要及遷移指南

### DISCUSSION THREAD

**Discussion by syimyuzya** [nk2028/qieyun-js#28] [Announcements]
Qieyun.js 0.13 ~~目前已漸穩定，將在推導方案更新至少四套並測試後擇日發佈~~ 已於 2022-04-24 發佈。

代碼在 ~~[nk2028/qieyun-js 的 dev-0.13 分支](https://github.com/nk2028/qieyun-js/tree/dev-0.13)~~ main 分支下。

試用：可以在 https://syimyuzya.github.io/qieyun-autoderiver/ 的控制臺下立即試用（該 fork 同時包含若干已更新到 0.13 的推導方案可供試用），或可在 Node.js 下 ~~`npm i qieyun@syimyuzya/qieyun-js#dist-dev` 試用~~ `npm i qieyun@latest` 開始使用。

文檔：~~暫未部署，可以直接看代碼中的文檔註釋，或者 git clone 下來用 `python prepare/main.py && npm run doc:html` 生成~~ 見 https://nk2028.shn.hk/qieyun-js

## Qieyun.js API 更動概要

資料相關：

- `iter音韻地位` 與 `query字頭` 兩函數移至 `Qieyun.資料` 下
- 移除 `音韻地位` 的 `.反切` `.代表字` `.條目` 三項，功能由 `Qieyun.資料.query音韻地位` 取代
- `query字頭` 與 `query音韻地位` 返回的條目所含信息增加「反切」「韻目原貌」兩項

音韻地位相關：

- 放寬 `音韻地位` 建立時的限制，允許在建立時「呼」和「重紐」不為 `null`（無論何種聲韻組合）
  - 當然，仍然不允許在不應為 `null` 的場合下為 `null`
  - 編碼向上兼容 0.12（只有新增的合法碼位，既有編碼無改動）
- 雖然創建時限制非常寬，但運用時往往需要更多限制以及統一等價音韻地位表達方式等（且標準寬嚴可能不惟一）
  - 該功能由 `Qieyun.適配分析體系` 擔任
  - 具體可參見其文檔註釋（在 [`src/lib/分析體系.ts`](https://github.com/nk2028/qieyun-js/blob/dev-0.13/src/lib/%E5%88%86%E6%9E%90%E9%AB%94%E7%B3%BB.ts)）及單元測試（在 [`src/lib/分析體系.spec.ts`](https://github.com/nk2028/qieyun-js/blob/dev-0.13/src/lib/%E5%88%86%E6%9E%90%E9%AB%94%E7%B3%BB.spec.ts)）
  - 關於 Qieyun.js 0.13 預設體系（"v2" 與 "v2ext" 兩種），亦可參見下文「遷移指南」中「音韻地位差異」部分
- 新增 `Qieyun.表達式` 裡面含有一些預定義的常用表達式

autoderiver 與 qieyun-examples 相關：

- `Qieyun.推導方案.建立` 為 autoderiver 和 qieyun-examples 兩者提供統一的函數包裝及解析選項列表等功能
  - 詳見[該模塊代碼](https://github.com/nk2028/qieyun-js/blob/dev-0.13/src/lib/%E6%8E%A8%E5%B0%8E%E6%96%B9%E6%A1%88.ts)
- 由其所產生的推導函數，自帶音韻地位合法性檢查（依 v2ext 體系，因此推導方案代碼只需考慮符合 v2ext 格式的音韻地位即可）

  - 若推導方案需要更嚴格的檢查或正則化，可在代碼中調用 `Qieyun.適配分析體系` 整體調整，或用 `音韻地位.調整` 來精細調整
  - 推導方案中的用例可參見新版的[切韻拼音](https://github.com/nk2028/qieyun-examples/blob/dev-qieyun-0.13/tupa.js)及[有女羅馬字](https://github.com/nk2028/qieyun-examples/blob/dev-qieyun-0.13/blankego.js)推導代碼開頭部分（在 qieyun-examples 的 dev-qieyun-0.13 分支）
- 舊推導代碼可以啟用「兼容模式」在 Qieyun.js 0.13 下繼續使用一段時間

  - 該功能只會保留一段時間（也不會加入文檔），請儘快依推導方案升級指南來更新代碼
  - 啟用「兼容模式」後，傳給包裝後推導函數的音韻地位，會在繼續傳給方案代碼前，自動轉換為符合舊版資料格式的音韻地位
  - 啟用「兼容模式」方法是在選項列表中加入 `['$legacy', true]` 這樣一項，`Qieyun.推導方案.建立` 會識別該項，且該項不會包含在解析後的選項列表中（qieyun-examples 的 dev-qieyun-0.13 分支下當未升級的代碼均已做該處理）

### qieyun-autoderiver 特別注意

- 建議使用 `Qieyun.推導方案.建立` 包裝原始推導函數以及解析選選列表（以便和 qieyun-examples 統一邏輯）
  - 用該方法得到的選項列表，會去除無效項，且 `Array` 型的選項，其 `[0]` 已經是默認項的值而非下標。

- 請給推導方案提供訪問 `Qieyun` 的能力，方法是：

  ```js
  import * as Qieyun from 'qieyun'; // 不要用 import Qieyun，webpack 下會莫名得到 undefined
  global.Qieyun = Qieyun;
  ```

  【當然，也有其他方法比如 `const tmpFunc = new Function('Qieyun', '音韻地位', '字頭', '選項', code);` 然後 `Qieyun.推導方案.建立((...args) => tmpFunc(Qieyun, ...args))`，但用 global 的話還有個好處是能讓 `Qieyun` 在控制臺使用，因此建議用 global 方式

- 推導方案代碼、預設文章這兩者均有配套更新，舊版內容不能正常運作於 0.13，請使用 qieyun-examples ~~的 dev-qieyun-0.13~~ 最新的 main 分支（尚未部署於騰訊雲）及 qieyun-text-label ~~的 v2~~ 最新的 main 分支下的內容

### 推導代碼

- 如果想在代碼中使用 `` is`...` `` 格式的 `音韻地位.屬於`，請將其定義為 `const is = (...x) => 音韻地位.屬於(...x)`，注意請使用 `...x`

### 音韻地位差異

此處列出 Qieyun.js 0.13 預設資料所用音韻地位體系（稱為 "v2" 體系）與舊版的差異。

v2 調整：

- 清韻不再標重紐（舊版清B「碧」小韻現為庚三）
  - 例：「辟」幫清入、「碧」幫三庚入

v2（及預設資料）新增情形：

- 端組麻三、庚二、佳韻：資料有「爹」「打」「箉」（定開佳上）三小韻（「地」是舊資料已經有的）
  - 例：「打」端開二庚上
- 祭廢韻平上聲：資料聲母有章組、云以日母，《廣韻》寄於齊灰咍韻
  - 例：「茝」昌開廢上
- 蒸B、幽B、陽A（注意與之對立的是 `重紐 = null`，並沒有蒸A幽A陽B）：資料蒸B、幽B有牙喉音，陽A僅脣音「𩦠」一小韻
  - 例：「抑」影開B蒸入、「憶」影開蒸入、「𩦠」並A陽上
- 匣母三等：於重紐韻須分AB；資料僅「礥」小韻一筆
  - 例：「礥」匣開A眞平
- 歌三脣音：注意其（雖然不屬於輕脣十韻但）當為輕脣，資料有「縛」小韻一筆
  - 例：「縛」幫三歌去

v2 取消的情形：

（凡說明「無統一校法」者，`適配分析體系.v2` 會拒絕處理該地位；另有 v2lenient 體系可用，會對該類地位保持原樣，如需處理請注意覆蓋這些情況）

（注「僅資料」者，指 v2 正則化規則中本身沒有直接禁止該類地位，也不會轉換該類地位，但資料中這類地位都被校了）

- 章組云以日母非三等：二等章組、日母分別校為莊組、孃母，其餘除蟹攝齊灰咍韻（調整為祭廢韻）外無統一校法
- 精組二等：校為莊組
- 莊組一四等：無統一校法
- 莊組眞欣韻：校為臻韻
- 羣、邪母非三等（僅資料）：邪校為從，羣母無統一校法
- 歌韻三等銳音（僅資料）：校為一等

v2ext 比 v2 額外多出組合：

（處理預設資料時不會遇到，但若想處理其他來源資料，可能需要注意以下額外多出的聲韻組合與對立）

- 寒歌韻脣音可能有「呼」為「開」者
  - 例：「爸」可能為「幫開一歌平」（v2：幫一歌平）
- 咍、嚴韻可能有脣音，凡韻可能有非脣音
  - 例：「劎」可能為「見凡去」（v2：見嚴去）
- 清韻的分重紐聲母可能有「B類」（但不會有明顯標出A類的）
  - 例：「碧」可能為「幫B清入」或「幫開B清入」（v2：幫三庚入）
- 除「地」「爹」「打」「箉」外可能有其他端組二三等或知組一四等
- 章組云以日母可能拼齊、灰、咍韻
  - 例：「茝」可能為「昌咍上」（v2：昌開廢上）
- 可能有精組二等、莊組一四等、章組以日母非三等
- 可能有云母非三等
- 可能有莊組眞欣韻

**Comment by syimyuzya** [comment]
@ayaka14732 @graphemecluster @untunt beta 版可以試用了～

**Comment by graphemecluster** [comment]
> 請給推導方案提供訪問 `Qieyun` 的能力

收到！

> webpack 下會莫名得到 undefined

下一版的 qieyun-autoderiver 會改用 Parcel ～

> 請將其定義為 `const is = (...x) => 音韻地位.屬於(...x)`

另建議同時定義 `const when = (...x) => 音韻地位.判斷(...x)`

**Reply by syimyuzya** [reply]
Parcel 好耶～

**Reply by syimyuzya** [reply]
對啦～隨着 Qieyun.js 0.13.0 發佈，現在 qieyun-text-label 和 qieyun-examples 最新的 main 分支內容已經是對應 0.13 的了。（但因為舊版推導器仍然在線，騰訊雲的 qieyun-examples 正暫停更新，目前仍是舊版（因為新舊資料不互通），更新 qieyun-autoderiver 的鏈接時注意一下這一點～
